pkg/controller/velerobackup/hiveobject_test.go

```diff
@@ -0,0 +1,185 @@
+			hiveObject:      ro2ho(testclusterdeployment.Build(unchangedClusterDeploymentBase())),
+			expectedChanged: false,
+			checksumFunc:    controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:            "Changed ClusterDeployment",
+			hiveObject:      ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase())),
+			expectedChanged: true,
+			checksumFunc:    controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:            "Unchanged SyncSet",
+			hiveObject:      ro2ho(testsyncset.Build(unchangedSyncSetBase())),
+			expectedChanged: false,
+			checksumFunc:    controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:            "Changed SyncSet",
+			hiveObject:      ro2ho(testsyncset.Build(changedSyncSetBase())),
+			expectedChanged: true,
+			checksumFunc:    controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:            "Error from Checksum func",
+			hiveObject:      ro2ho(testsyncset.Build(unchangedSyncSetBase())),
+			expectedChanged: false,
+			checksumFunc:    getHashOfObjectsFuncReturnsError,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Arrange
+
+			// Act
+			actualChanged := test.hiveObject.hasChanged()
+
+			// Assert
+			assert.Equal(t, test.expectedChanged, actualChanged)
+		})
+	}
+}
+
+func TestCalculateChecksum(t *testing.T) {
+	tests := []struct {
+		name             string
+		checksumObject   *hivev1.ClusterDeployment
+		expectedChecksum string
+		checksumFunc     controllerutils.ChecksumOfObjectsFunc
+	}{
+		{
+			name:             "Valid ClusterDeployment Checksum",
+			checksumObject:   testclusterdeployment.Build(unchangedClusterDeploymentBase()),
+			expectedChecksum: testclusterdeployment.Build(unchangedClusterDeploymentBase()).Annotations[controllerutils.LastBackupAnnotation],
+			checksumFunc:     controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:             "Error from Checksum func",
+			checksumObject:   testclusterdeployment.Build(unchangedClusterDeploymentBase()),
+			expectedChecksum: errChecksum,
+			checksumFunc:     getHashOfObjectsFuncReturnsError,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Arrange
+
+			// Act
+			actualChecksum := calculateChecksum(test.checksumObject.ObjectMeta, test.checksumObject.Spec, test.checksumFunc, log.WithField("controller", controllerName))
+
+			// Assert
+			assert.Equal(t, test.expectedChecksum, actualChecksum)
+		})
+	}
+}
+
+func TestSetChecksum(t *testing.T) {
+	tests := []struct {
+		name             string
+		hiveObject       *hiveObject
+		expectedChecksum string
+	}{
+		{
+			name:             "Valid ClusterDeployment Checksum",
+			hiveObject:       ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase())),
+			expectedChecksum: testclusterdeployment.Build(changedClusterDeploymentBase(), testclusterdeployment.WithBackupChecksum(defaultChecksumFunc)).Annotations[controllerutils.LastBackupAnnotation],
```

Comment:
This is too complicated for what you are trying to test. Instead, do something like this. Note that you do not need to test each different type individually as setting the checksum does not depend upon the type.
```
func TestSetChecksum(t *testing.T) {
  ro := &hivev1.ClusterDeployment{
    ObjectMeta: metav1.ObjectMeta{
      Annotations: map[string]string{controllerutils.LastBackupAnnotation: "old-checksum"},
    },
  }
  hiveObject := &hiveObject{
    object: ro,
    checksum: "new-checksum"
  }
  hiveObject.setChecksumAnnotation()
  actualChecksum := ro.Annotations[controllerutils.LastBackupAnnotation]
  assert.Equals(t, "new-checksum", actualChecksum, "unexpected checksum")
}
```